flang/lib/Lower/CallInterface.cpp

```diff
@@ -0,0 +1,455 @@
+
+#include "flang/Lower/CallInterface.h"
+#include "flang/Evaluate/characteristics.h"
+#include "flang/Evaluate/fold.h"
+#include "flang/Lower/Bridge.h"
+#include "flang/Lower/FIRBuilder.h"
+#include "flang/Lower/PFTBuilder.h"
+#include "flang/Optimizer/Dialect/FIRDialect.h"
+#include "flang/Optimizer/Support/InternalNames.h"
+#include "flang/Semantics/symbol.h"
+#include "flang/Semantics/tools.h"
+
+//===----------------------------------------------------------------------===//
+// Caller side interface implementation
+//===----------------------------------------------------------------------===//
+
+bool Fortran::lower::CallerInterface::hasAlternateReturns() const {
+  return procRef.HasAlternateReturns();
+}
+
+std::string Fortran::lower::CallerInterface::getMangledName() const {
+  const auto &proc = procRef.proc();
+  if (const auto *symbol = proc.GetSymbol())
+    return converter.mangleName(*symbol);
+  assert(proc.GetSpecificIntrinsic() &&
+         "expected intrinsic procedure in designator");
+  return proc.GetName();
+}
+
+mlir::Location Fortran::lower::CallerInterface::getCalleeLocation() const {
+  const auto &proc = procRef.proc();
+  if (const auto *symbol = proc.GetSymbol()) {
+    // FIXME: If the callee is defined in the same file but after the current
+    // unit we do cannot get its location here and the funcOp is created at the
+    // wrong location (i.e, the caller location).
+    if (const auto *details =
+            symbol->detailsIf<Fortran::semantics::ProcEntityDetails>())
+      if (const auto *interfaceSymbol = details->interface().symbol())
+        symbol = interfaceSymbol;
+    return converter.genLocation(symbol->name());
+  }
+  // Unknown location for intrinsics.
+  return converter.genLocation();
+}
+
+Fortran::evaluate::characteristics::Procedure
+Fortran::lower::CallerInterface::characterize() const {
+  auto &foldingContext = converter.getFoldingContext();
+  auto characteristic =
+      Fortran::evaluate::characteristics::Procedure::Characterize(
+          procRef.proc(), foldingContext.intrinsics());
+  assert(characteristic && "Failed to get characteristic from procRef");
+  // The characteristic may not contain the argument characteristic if no
+  // the ProcedureDesignator has no interface.
+  if (!characteristic->HasExplicitInterface()) {
+    for (const auto &arg : procRef.arguments()) {
+      // Argument cannot be optional with implicit interface
+      const auto *expr = arg.value().UnwrapExpr();
+      assert(expr &&
+             "argument in call with implicit interface cannot be assumed type");
+      auto argCharacteristic =
+          Fortran::evaluate::characteristics::DummyArgument::FromActual(
+              "actual", *expr, foldingContext);
+      assert(argCharacteristic &&
+             "failed to characterize argument in implicit call");
+      characteristic->dummyArguments.emplace_back(
+          std::move(*argCharacteristic));
+    }
```

Comment:
The call Fortran::evaluate::characteristics::Procedure::Characterize(procRef.proc()) is doing this for us since the result type is always explicit (its the type of the function symbol).